Fix specificity issue between theme and user styles

oandregal opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/29381

## How to test

- Use a theme with support for FSE (templates) and Global Style (theme.json), such as TT1-blocks.
- Via theme.json target the post title for color styles by adding:

```json
"core/post-title/h2": {
    "color": {
        "text": "green",
        "background": "yellow"
    }
}
```

- Go to the editor, add three post title blocks:
  - Select text & background preset colors (not custom) for the first.
  - Select text & background custom colors (not preset) for the second.
  - Leave the third untouched.
- Publish the post and go to the front end.

The expected behavior would be that the user choices are respected, both in the editor and front-end.

youknowriad commented on lib/class-wp-theme-json.php:
```diff
@@ -1003,6 +1003,47 @@ private function get_block_styles() {
+			// such as `.has-text-color .has-red-color`, etc.
+			//
+			// In those cases, we want the user preferences
+			// to overrule the theme's, hence this negations.
```
This sounds very fragile to me and not exhaustive, what do you mean by "user preferences" here?

nosolosw replied:
The issue here is that preset classes are enqueued in the global stylesheet as:

`.has-red-color { color: red; }`

while the post title block (and potentially others now that block supports can be used to target inner elements) can do:

`h2.wp-block-post-title { color: "theme-color"; } `

Hence, the theme setting specificity won't let the user preference to take effect (user selects a preset color for the block => the class gets attached to the post content).

Alternatives I've considered:

1. Bumping up the specificity of the preset classes => not robust, we can't be sure what's the higher block specificity we have to counteract.

2. Add more block selectors to the preset classes rulesets, so we'd end up with:

```css
h2.wp-block-post-title.has-red-color,
<other-selectors>,
.has-red-color { 
  color: red
}
```

To avoid having to add all block selectors for all preset classes, which is a lot of CSS, I think we should output only the selectors for blocks whose corresponding property is styled (if they have a color => add selector to list).

3. Finally, I thought about the current approach: check for the presence of `has-text-color`, etc. It outputs a lot less CSS. I think it addresses all the cases we have. The only thing I'm not 100% comfortable with is the font-size check, as we only output `.has-<value>-font-size` in those cases so the attribute selector can only check for `-font-size`. Theoretically, I understand this bit may report false positives, but I'm not sure it'll be an issue in practice.

I have a slight preference for 3 given the amount of CSS we remove from shipping, but 2 would also work.

youknowriad replied:
Don't you think when a preset is applied, it should always be favored? in which case, adding `important` (forget the dogma) to the generated styles could be actually the best way to go here?

youknowriad replied:
I'm saying that because I don't believe relying on `has-text-color` and similar classes is a good thing, it adds an indirect dependency to something the block might or might not apply. (the block supports should be seen as just a shortcut for block authors and not something mandatory when using presets...)

nosolosw replied:
Ah, I see your point now: it's about the fact that `has-text-color` classes are only added by _core's UI controls_ and not necessarily by any other custom UI controls. :ok_hand:

Actually. Given how the style system works, `!important` is a good choice for preset classes the user attaches to the post content. If those values need to be overridden, in addition to regular CSS, we can offer hooks for the theme.json processing, so plugins have an alternate mechanism to modify this behavior.

Thanks for this feedback!

nosolosw replied:
https://github.com/WordPress/gutenberg/pull/29533/commits/980a0b6c981970388430d374cbe5d017c865c29c

oandregal committed:
Add test cases

oandregal committed:
Make preset values !important

Given how the style system works, !important is a good choice
for preset classes the user attaches to the post content.
If those values need to be overridden, in addition to regular CSS,
we can offer hooks for the theme.json processing,
so plugins have an alternate mechanism to modify this behavior.

oandregal committed:
Rename test

youknowriad reviewed (approved):
Makes sense to me, but other opinions would be good too.

aristath reviewed (approved):
LGTM 👍 
I'm sure lots of people will say that `!important` is evil, but sometimes it's the best tool for the job.

oandregal committed:
Make linter happy

nosolosw commented:
OK, I think I'm going to land this so we can get better feedback ― this is something that's easily reversible so we can course-correct later, if necessary.

oandregal commented:
Wanted to share https://github.com/WordPress/gutenberg/issues/29985#issuecomment-814726857 for all that follow this PR.

webmandesign commented:
Hi guys,

I'm very late to the party, but the issue with `!important` in recent WP5.9 led me here.
I have to voice one of my worries, that I can see being very real reading your comments here (but is not related to `!important`).

From your [conversation above](#discussion_r587288793):

> @youknowriad : I'm saying that because I don't believe relying on `has-text-color` and similar classes is a good thing, it adds an indirect dependency to something the block might or might not apply. (the block supports should be seen as just a shortcut for block authors and not something mandatory when using presets...)

> @oandregal : Ah, I see your point now: it's about the fact that `has-text-color` classes are only added by core's UI controls and not necessarily by any other custom UI controls.

There may be potentially a big issue with Gutenberg **not** applying "BEMish" naming on modifier CSS classes. Although it uses BEM for block classes (such as `.wp-block-button` &rarr; `.wp-block-button__link`), it fails with modifier classes such as `.has-red-color` _(I will use a text color modifier class as an example in here, but the issue applies for all modifier classes)_.

It is especially impossible to target `.has-red-color`, `.has-black-color` and `.has-white-color` classes at once for any batch styling without targeting a helper `.has-text-color` class. If `.has-text-color` class is not something a theme (and plugins) can rely on, then we have a problem.

If Gutenberg used BEMish naming for modifier classes such as `.has-<PROPERTY>--<VALUE>`, we would get `.has-color--red` or `.has-font-size--small` and theme authors could target these with `[class*="has-<PROPERTY>--"]` selectors. There would be no need for helper classes such as `.has-text-color` then.

Also, there is no way for theme authors to know about potential redundancy of `.has-text-color` class and I'd say it is used quite often in the wild. That's why theme authors rely on it (and similar classes).

It seems like there is no documentation for Gutenberg best practices and standards so even extension plugins can follow these to make the ecosystem more consistent. (&rarr; could be of an interest for @annezazu re dev focused communication site)

It's also confusing that colors have a helper class (`.has-text-color`, `.has-background` and `.has-background-gradient`), but font size does not, for example.

But like I've mentioned, this issue could be resolved with more modular naming of CSS classes.

Do you think it is worth to log a new issue ticket for this problem?

oandregal commented:
Hey Oliver. Sure, if there's an issue with how current classes work that merits an independent conversation.

webmandesign commented:
Hi André,

There is actually no issue currently with the classes.
Just reading your comments here brings up concerns about class naming and helper class redundancy.

I will open a new issue for discussion. Thanks.

oandregal commented:
Hi Oliver, if there's no current issue but you still want to share your thoughts on something, I believe https://github.com/WordPress/gutenberg/discussions would be a better place as it helps keep Gutenberg issues for triaging actual product bugs/features/etc.

webmandesign commented:
Thanks for pointing me in the right direction.
